### PR TITLE
CALCITE-3158: Avatica: migrate build system to Gradle

### PR DESCRIPTION
**Pull request by vlsi**
https://issues.apache.org/jira/browse/CALCITE-3158

"Are we there yet?" checklist (in no particular order)
- [x] Build jars
- [x] Run tests
- [x] Checkstyle
- [x] Forbiddenapis
- [x] RAT
- [x] Load project to IDE
- [x] Automate license file generation
- [x] SpotBugs
- [ ] Docker images
- [x] generate POM files
- [x] centralize version management for dependencies (~ Maven's `<dependencyManagement>`) 
- [x] OWASP dependency check plugin
- [ ] "unused dependency"
- [ ] "used but undeclared dependency"
- [ ] errorprone
- [x] javadoc
- [x] javadoc in Java9+ (HTML5, link to https://docs.oracle.com/javase/9/docs/api/)
- [x] create release artifacts (avatica-src.tar.tz?)
- [x] publish to Nexus
- [x] publish to dist.apache.org
- [ ] fix Travis/Appveyor (current PR tests one Java version only to save time)
- [x] fix documentation/site (Maven -> Gradle)

Note: this PR uses `com.github.vlsi.*` plugins, and those are pending approval for Gradle Plugin Portal.

In the mean time the workaround is to clone the plugins side by side with calcite-avatica repository (and it will work transparently thanks to `settings.gradle.kts`).

In other words:
1) `git clone https://github.com/apache/calcite-avatica`
2) `git clone --recurse-submodules https://github.com/vlsi/vlsi-release-plugins`
3) `cd calcite-avatica && ./gradlew build`

### CODE REVIEW

**Comment by vlsi** [comment]
@risdenk , @joshelser , it looks like `protoc` can be downloaded from Maven Central.
Does that mean we should remove `src/main/java/org/apache/calcite/avatica/proto/*.java` files?

See the following in `core/build.gradle.kts`:

```
protobuf {
    protoc {
        // Download from repositories
        artifact = "com.google.protobuf:protoc:3.6.1"
    }
}
```

**Comment by vlsi** [comment]
I see current build heavily relies on Docker.
Does that mean "Docker" is the only valid way to build Avatica?

It does not sound quite right.

**Comment by vlsi** [comment]
Issue for Hamcrest licensing info: https://github.com/hamcrest/JavaHamcrest/issues/264

**Comment by vlsi** [comment]
Issue for hsqldb licensing info: https://sourceforge.net/p/hsqldb/feature-requests/351/

**Comment by zabetak** [comment]
Yes, I was referring to building SNAPSHOTs. I tested again and now they can be built without signing.

I tried `-PuseGpgCmd` and I am pretty happy with it. It is even better than typing. Thanks!

Regarding Q3, basically I was wondering how can I simulate a full release cycle without committing anything to the official GitHub/GitBox repo or [Nexus](https://repository.apache.org/). My idea was to use [asflike-release-environment](https://github.com/vlsi/asflike-release-environment) as you mentioned in the emails.

**Comment by zabetak** [comment]
It would be nice to have somewhere in the doc a few instructions on which of these properties can be configured and how; how can we override the defaults if necessary. I spend ~10 minutes and didn't manage to understand how can this be done by `gradle.properties`. 

Other than that it seems that `asflike-release-environment` might need some tweaking. After starting the container as you indicated I am running the following command:

`gradle prepareVote -PasfTestGitSourceUsername=test -PasfTestGitSourcePassword=test -Prc=0 -PuseGpgCmd`

and I get the following output. From the error I assume that the svn-server is missing some info.

```
> Configure project :
Building Apache Calcite Avatica 1.16.0

> Task :pushRcTag
Pushing tag to Git remote release-origin-test: git://127.0.0.1/calcite.git
Message from release-origin-test: 
Updated remote ref Ref[refs/heads/asf-site=4b66da1846d5efdf1665cb29439a44a0174dc5a0(-1)]: null
Updated remote ref Ref[refs/heads/gh-pages=4b66da1846d5efdf1665cb29439a44a0174dc5a0(-1)]: null
Updated remote ref Ref[refs/heads/master=4b66da1846d5efdf1665cb29439a44a0174dc5a0(-1)]: null
Updated remote ref Ref[refs/tags/avatica-1.16.0-rc0=12708c58848f7a7d63d86c43b70a58caf429c415(-1)]: RemoteRefUpdate[remoteName=refs/tags/avatica-1.16.0-rc0, UP_TO_DATE, (null)...12708c58848f7a7d63d86c43b70a58caf429c415, srcRef=refs/tags/avatica-1.16.0-rc0, message=null]

> Task :noop-driver:initializeNexusStagingRepository
Initialized stagingRepositoryId orgapachecalcite-1570723739049 for repository nexus

> Task :stageSvnDist FAILED
Unable to check existence of http://127.0.0.1/svn/dist/dev. Error: svn: E170013: Unable to connect to a repository at URL 'http://127.0.0.1/svn/dist/dev'
svn: E120108: Error running context: The server unexpectedly closed the connection.

Unable to check existence of http://127.0.0.1/svn/dist/dev/calcite. Error: svn: E170013: Unable to connect to a repository at URL 'http://127.0.0.1/svn/dist/dev/calcite'
svn: E120108: Error running context: The server unexpectedly closed the connection.

Unable to check existence of http://127.0.0.1/svn/dist/dev/calcite/apache-calcite-avatica-1.16.0-rc0. Error: svn: E170013: Unable to connect to a repository at URL 'http://127.0.0.1/svn/dist/dev/calcite/apache-calcite-avatica-1.16.0-rc0'
svn: E120108: Error running context: The server unexpectedly closed the connection.

Svnmucc: executing mkdir
dev
mkdir
dev/calcite
mkdir
dev/calcite/apache-calcite-avatica-1.16.0-rc0
svnmucc: E170013: Unable to connect to a repository at URL 'http://127.0.0.1/svn/dist'
svnmucc: E120108: Error running context: The server unexpectedly closed the connection.
```

**Comment by vlsi** [comment]
>I spend ~10 minutes and didn't manage to understand how can this be done by gradle.properties

Isn't that nice?
How much time do you need to figure out the release with Maven?

The suggested location of the properties is `~/.gradle/gradle.properties`

https://docs.gradle.org/current/userguide/build_environment.html#sec:gradle_configuration_properties :

>gradle.properties in GRADLE_USER_HOME directory.
>GRADLE_USER_HOME
>  Specifies the Gradle user home directory (which defaults to $USER_HOME/.gradle if not set).

> and I get the following output. From the error I assume that the svn-server is missing some info.

Can you please clarify how did you create the svn-server?
./recreate.sh calcite ?

Right after the recreate it should spin up SVN server at http://localhost/svn, and it should contain folder http://localhost/svn/dist/dev/calcite/

Does it work?
Does http://127.0.0.1/svn work?
It looks like the default URLs use 127.0.0.1 rather than localhost.

**Comment by julianhyde** [comment]
A different idea is to add this to your `maven-release-plugin` properties in `pom.xml`:
```
        <configuration>
          <localCheckout>true</localCheckout>
          <pushChanges>false</pushChanges>
        </configuration>
```

It allows a release to be made (or prepared) without pushing to the remote git repository. (This is for maven, obviously. I don't know whether there's an equivalent for gradle.)

It will serve your purposes, and we could even make it the default.

**Comment by vlsi** [comment]
> It allows a release to be made (or prepared) without pushing to the remote git repository.

@julianhyde , please show me how Maven enables to stage and promote releases via https://dist.apache.org

**Comment by julianhyde** [comment]
@vlsi Stamatis asked "how can I simulate a full release cycle without committing anything to the official GitHub/GitBox repo" and I was answering that question. I don't have any advice about nexus or dist.apache.org.

**Comment by vlsi** [comment]
I see.
The thing is Gradle-based build enables to simulate a full release cycle without committing anything to the official GitHub/GitBox repo. It spins the local Git/Nexus/SVN servers.

**Comment by zabetak** [comment]
@vlsi thanks for taking the time to help, much appreciated. 
 
Regarding the properties what I would to know how I can point to a different Nexus URL (other than the default http://localhost:8080/). Sorry for not being clear.

Regarding _asflike-release-environment_ what I am doing is the following:
`sudo ./recreate.sh calcite`

The docker ps commands shows the following:

```
CONTAINER ID        IMAGE                           COMMAND                  CREATED             STATUS              PORTS                                                                                   NAMES
b075163dbaa4        vlsi/nexus-stub                 "java -Xms128m -cp /…"   30 seconds ago      Up 28 seconds       0.0.0.0:8080->8080/tcp                                                                  nexus-stub
ea88359e4bb8        asflikereleaseenvironment_vcs   "/init"                  30 seconds ago      Up 29 seconds       443/tcp, 0.0.0.0:80->80/tcp, 0.0.0.0:3960->3960/tcp, 0.0.0.0:9418->9418/tcp, 3690/tcp   vcs-server
```

The svn-server is up and running but after putting the credentials (username:`test`, password:`test`) I get the [following error](https://ibb.co/NTQ3c59) no matter if I use localhost or 127.0.0.1. I will try to figure out what happens tomorrow.

@julianhyde Thanks for the tip about maven.

**Comment by vlsi** [comment]
> Regarding the properties what I would to know how I can point to a different Nexus URL (other than the default http://localhost:8080/).

There's no property for that yet. The following should work:

```kotlin
releaseParams {
  nexus {
    url.set("https://repository.apache.org")
  }
}
```

> The docker ps commands shows the following:

It shows the same for me. Do you probably have another server on port 80?

**Comment by zabetak** [comment]
The connection to the svn server works fine via the `svn://` protocol. However, the connection via `http://` still does not work (for me). In any case it is not a problem of _asflike-release-environment_ since I tried using [elleforio/svn-server](https://hub.docker.com/r/elleflorio/svn-server/) directly and I still encounter the same issue. 

Fun fact: Before creating any repository the connection via http is working fine and it shows me the empty folder but just after creating a new repo I get the error mentioned earlier.

Anyways given that issue is not directly related with this PR I will stop the discussion here since I guess it is not interesting for Calcite people.

**Comment by vlsi** [comment]
Frankly speaking, having a dummy environment so everybody can try making a release without pushing to the official repositories was one of the goals.

I've added readme for stage-vote-release-plugin.

https://github.com/vlsi/vlsi-release-plugins/tree/master/plugins/stage-vote-release-plugin

Of course, I agree asflike-release-environment is not a top priority, so it should not block the review / merge.

**Comment by zabetak** [comment]
After @vlsi [commit](https://github.com/vlsi/asflike-release-environment/commit/4c3ed6d42452a578439b994761582b70bf426c9d) in _asflike-release-environment_ the afforementioned problems with svn and http are resolved. 

Now I tested again `gradle prepareVote -Prc=0 -PuseGpgCmd` and I get the following errors:
```
...
> Task :metrics:publishMetricsPublicationToNexusRepository FAILED
> Task :noop-driver:publishNoop-driverPublicationToNexusRepository FAILED
> Task :metrics-dropwizardmetrics:publishMetrics-dropwizardmetricsPublicationToNexusRepository FAILED
> Task :server:publishServerPublicationToNexusRepository FAILED
> Task :tck:publishTckPublicationToNexusRepository FAILED
> Task :shaded:avatica:publishAvaticaPublicationToNexusRepository FAILED
> Task :core:publishCorePublicationToNexusRepository FAILED
> Task :standalone-server:publishStandalone-serverPublicationToNexusRepository FAILED

FAILURE: Build completed with 8 failures.

1: Task failed with an exception.
-----------
* What went wrong:
Execution failed for task ':metrics:publishMetricsPublicationToNexusRepository'.
> Failed to publish publication 'metrics' to repository 'nexus'
   > java.io.EOFException: input contained no data

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.
==============================================================================

2: Task failed with an exception.
-----------
* What went wrong:
Execution failed for task ':noop-driver:publishNoop-driverPublicationToNexusRepository'.
> Failed to publish publication 'noop-driver' to repository 'nexus'
   > java.io.EOFException: input contained no data

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.
==============================================================================

3: Task failed with an exception.
-----------
* What went wrong:
Execution failed for task ':metrics-dropwizardmetrics:publishMetrics-dropwizardmetricsPublicationToNexusRepository'.
> Failed to publish publication 'metrics-dropwizardmetrics' to repository 'nexus'
   > java.io.EOFException: input contained no data

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.
==============================================================================

4: Task failed with an exception.
-----------
* What went wrong:
Execution failed for task ':server:publishServerPublicationToNexusRepository'.
> Failed to publish publication 'server' to repository 'nexus'
   > java.io.EOFException: input contained no data

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.
==============================================================================

5: Task failed with an exception.
-----------
* What went wrong:
Execution failed for task ':tck:publishTckPublicationToNexusRepository'.
> Failed to publish publication 'tck' to repository 'nexus'
   > java.io.EOFException: input contained no data

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.
==============================================================================

6: Task failed with an exception.
-----------
* What went wrong:
Execution failed for task ':shaded:avatica:publishAvaticaPublicationToNexusRepository'.
> Failed to publish publication 'avatica' to repository 'nexus'
   > java.io.EOFException: input contained no data

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.
==============================================================================

7: Task failed with an exception.
-----------
* What went wrong:
Execution failed for task ':core:publishCorePublicationToNexusRepository'.
> Failed to publish publication 'core' to repository 'nexus'
   > java.io.EOFException: input contained no data

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.
==============================================================================

8: Task failed with an exception.
-----------
* What went wrong:
Execution failed for task ':standalone-server:publishStandalone-serverPublicationToNexusRepository'.
> Failed to publish publication 'standalone-server' to repository 'nexus'
   > java.io.EOFException: input contained no data
```

**Comment by vlsi** [comment]
@zabetak , can you please add --stacktrace and show the stacktrace?

Did you use gradle or gradlew?

**Comment by vlsi** [comment]
The thing there is the current `nexus-stub` always returns empty file for `maven-metadata.xml`, and it confuses Gradle 5.6+ (see https://github.com/vlsi/nexus-stub/issues/1 )

**Comment by zabetak** [comment]
Indeed I was using Gradle 5.6.2. I switched to 5.5.1 and prepareVote finishes normally. Success :) 
@vlsi I will play a bit more and keep you posted if I find something else!

**Comment by vlsi** [comment]
I've implemented `nexus-stub` a bit, so `asflike-release-environment` can now be used with Gradle 5.6.

**Comment by zabetak** [comment]
Is it normal that `prepareVote` task does not create a release/build directory? 

After successfully completing the task I don't see the apache-calcite-avatica-* artifacts. 
I tried:
`avatica> find . -name "apache-calcite-avatica*"` and it does not print anything.

If I do `gradle build` instead I see the artifacts under release/build/distributions directory.

**Comment by vlsi** [comment]
> Is it normal that prepareVote task does not create a release/build directory?

Of course, it is not. I've fixed that.

**Comment by zabetak** [comment]
There seems to be a problem with the staging repo. The link (http://127.0.0.1:8080/content/repositories/orgapachecalcite-1571224925812/org/apache/calcite/) mentioned in the draft VOTE email points to  `Whitelabel Error Page`.

**Comment by vlsi** [comment]
@zabetak , well. `nexus-stub` does not support repository browse yet (https://github.com/vlsi/nexus-stub/issues/2).

Please have a look at Apache JMeter 5.2-rc3 vote thread: https://lists.apache.org/thread.html/051df3ba125b45bcac6cff251f139c770522f468c3520323b3d59255@%3Cdev.jmeter.apache.org%3E

All the links (including site and javadoc preview) in the mail are generated and populated with `prepareVote`

**Comment by F21** [comment]
I've been testing this out in docker and this is what I've got so far:

I first started the stub environment using `./recreate.sh calcite`.

After cloning your branch, I mounted it into a Gradle container in interactive mode:
```
docker run -v "$PWD":/home/gradle/project -v "/c/Users/Francis Chuang/AppData/Roaming/gnupg":/.gnupg -w /home/gradle/project --network="host" -it gradle sh
apt update
apt install gpg
mkdir -p /root/.gnupg
cp -r /.gnupg/ /root/
chmod -R 700 /root/.gnupg/
rm -rf /root/.gnupg/*.lock
gradle prepareVote -PasfTestGitSourceUsername=test -PasfTestGitSourcePassword=test -PallowUncommittedChanges -Prc=0 -PuseGpgCmd
```
Gpg shows that my keys are in the docker container:
```
# gpg --list-secret-keys
/root/.gnupg/pubring.kbx
------------------------
sec   rsa4096 2018-04-16 [SC]
      635665E0BE3F72552910CB74BBE44E923A970AB7
uid           [ultimate] Francis Chuang <francischuang@apache.org>
ssb   rsa4096 2018-04-16 [E]
```

The prepareVote tasks fails:
```
# gradle prepareVote -PasfTestGitSourceUsername=test -PasfTestGitSourcePassword=test -PallowUncommittedChanges -Prc=0 -PuseGpgCmd

> Configure project :
Building Apache Calcite Avatica 1.16.0

> Task :core:initializeNexusStagingRepository
Initialized stagingRepositoryId orgapachecalcite-1572996459686 for repository nexus

> Task :metrics:signMetricsPublication FAILED

> Task :pushRcTag
Pushing tag to Git remote release-origin-test: git://127.0.0.1/calcite.git
Message from release-origin-test:
Updated remote ref Ref[refs/heads/asf-site=a9d6941660e39fd39e2554562ec2518775e9aac4(-1)]: null
Updated remote ref Ref[refs/heads/gh-pages=a9d6941660e39fd39e2554562ec2518775e9aac4(-1)]: null
Updated remote ref Ref[refs/heads/master=a9d6941660e39fd39e2554562ec2518775e9aac4(-1)]: null
Updated remote ref Ref[refs/tags/avatica-1.16.0-rc0=acb8ec778af888048b259bcb95565f018b2f1591(-1)]: RemoteRefUpdate[remoteName=refs/tags/avatica-1.16.0-rc0, UP_TO_DATE, (null)...acb8ec778af888048b259bcb95565f018b2f1591, srcRef=refs/tags/avatica-1.16.0-rc0, message=null]

FAILURE: Build failed with an exception.

* What went wrong:
A problem was found with the configuration of task ':metrics:signMetricsPublication'.
> No value has been specified for property 'signatory.keyId'.

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

Deprecated Gradle features were used in this build, making it incompatible with Gradle 6.0.
Use '--warning-mode all' to show the individual deprecation warnings.
See https://docs.gradle.org/5.6.4/userguide/command_line_interface.html#sec:command_line_warnings

BUILD FAILED in 3m 55s
37 actionable tasks: 18 executed, 19 up-to-date
```

**Comment by vlsi** [comment]
@F21 , thanks for giving it a try.

Have you tried adding `signing.gnupg.keyName=...` Gradle property? (see https://github.com/gradle/gradle/issues/8657 )

**Comment by zabetak** [comment]
I performed a dryRun of the release with gradle and then back to back a dryRun of the release using mvn. I know it is a bit out of scope since mvn will be completely replaced but I noticed that in the second case there are the following problems when performing the RAT check:
```
cat target/rat.txt | grep "\!?????"
 !????? gradlew.bat
 !????? .ratignore
 !????? gradlew
 !????? release/build/licenses/getLicenses/license.xml
 !????? release/build/distributions/apache-calcite-avatica-1.16.0-src.tar.gz.asc
 !????? release/build/distributions/apache-calcite-avatica-1.16.0-src.zip.asc
 !????? release/build/distributions/apache-calcite-avatica-1.16.0-src.tar.gz.sha512
 !????? release/build/distributions/apache-calcite-avatica-1.16.0-src.zip.sha512
 !????? src/main/config/licenses/hsqldb/LICENSE-hsqldb
 !????? src/main/config/licenses/hsqldb/LICENSE-hypersonic
 !????? .gradle/vcs-1/gc.properties
 !????? .gradle/buildOutputCleanup/cache.properties
 !????? .gradle/5.5.1/gc.properties
```
I suppose that these come up because .ratignore is not used by maven right?

**Comment by F21** [comment]
@vlsi , thanks, I was able to move forward a little bit. Now, it's complaining about username being null:

``
> Task :metrics:publishMetricsPublicationToNexusRepository FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':metrics:publishMetricsPublicationToNexusRepository'.
> Failed to publish publication 'metrics' to repository 'nexus'
   > Username must not be null!

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

Deprecated Gradle features were used in this build, making it incompatible with Gradle 6.0.
Use '--warning-mode all' to show the individual deprecation warnings.
See https://docs.gradle.org/5.6.4/userguide/command_line_interface.html#sec:command_line_warnings

BUILD FAILED in 1s
52 actionable tasks: 20 executed, 32 up-to-date
```

Again, I am doing everything in a docker container in interactive mode. I also started the gpg agent and authenticated my key as I've found that the console will prevent the key from being entered properly as the build would be working in the background and outputting messages:

```
docker run -v "$PWD":/home/gradle/project -v "/c/Users/Francis Chuang/AppData/Roaming/gnupg":/.gnupg -w /home/gradle/project --network="host" -it gradle sh
apt update
apt install gnupg2
mkdir -p /root/.gnupg
cp -r /.gnupg/ /root/
chmod -R 700 /root/.gnupg/
rm -rf /root/.gnupg/*.lock
export GPG_TTY=/dev/console
touch /root/.gnupg/gpg-agent.conf
echo 'default-cache-ttl 10000' >> /root/.gnupg/gpg-agent.conf
echo 'max-cache-ttl 10000' >> /root/.gnupg/gpg-agent.conf
"test" | gpg2 --local-user 635665E0BE3F72552910CB74BBE44E923A970AB7 --output /dev/null --sign -
gradle prepareVote -PasfTestGitSourceUsername=test -PasfTestGitSourcePassword=test -PallowUncommittedChanges -Prc=0 -PuseGpgCmd -Psigning.gnupg.keyName=635665E0BE3F72552910CB74BBE44E923A970AB7
```

**Comment by zabetak** [comment]
Hey @F21 ,

You need to set more or less the following properties:
```
signing.gnupg.executable=gpg
signing.gnupg.keyName=FXXXXXXX

asfCommitterId=zabetak
asfTestNexusUsername=test
asfTestNexusPassword=test
asfTestGitSourceUsername=test
asfTestGitSourcePassword=test
asfTestGitSitePreviewUsername=test
asfTestGitSitePreviewPassword=test
asfTestSvnUsername=test
asfTestSvnPassword=test
```
At least this does the trick for me.

**Comment by vlsi** [comment]
>Username must not be null!

com.github.vlsi.stage-vote-release should probably be updated to 1.37.0, and it will print the missing properties before it starts executing the tasks.

>I suppose that these come up because .ratignore is not used by maven right?

I guess so.
By the way, I guess the most part of `.ratignore` should be moved to `.gitignore`.

In other words, I think `.gitignore` and `.gitattributes` should be the primary source of the information, then `.ratignore` could add extra patterns (e.g. we know we can't add a license to `site/favicon.ico`, so that file has to be in `.ratignore`)

**Comment by vlsi** [comment]
@F21 , I've updated the plugin versions, so it reports the missing properties now:

```
gradle prepareVote -PasfTestGitSourceUsername=test -PasfTestGitSourcePassword=test -PallowUncommittedChanges -Prc=0 -PuseGpgCmd

...
* What went wrong:
Please correct the following before proceeding with the release:
1) Property asfTestSvnUsername is not specified
2) Property asfTestSvnPassword is not specified
3) Property asfTestNexusUsername is not specified
4) Property asfTestNexusPassword is not specified
5) Please specify signing key id via signing.gnupg.keyName (see https://github.com/gradle/gradle/issues/8657)
```

PS I see you are using `gradle:latest` container, thus it might silently upgrade to next Gradle version which is probably unwanted.
You probably want to use the same Gradle version that is specified in the wrapper configuration. For instance: `gradle:5.6.2-jre8`

**Comment by vlsi** [comment]
I've removed `pom.xml` files, and updated the documentation.

I suggest someone (@F21 ?) tries the scripts without Docker first, then we decide what to do with `docker.sh`.

As for me, Docker helps for website generation since Ruby gems do not install on my machine.
However, the release itself is just two commands (`./gradlew prepareVote -Prc=1 -Pasf` and `./gradlew publishDist -Prc=1`), so `docker.sh` might probably be simplified.

**Comment by zabetak** [comment]
> By the way, I guess the most part of `.ratignore` should be moved to `.gitignore`.

Can the [apache-rat-plugin](https://creadur.apache.org/rat/apache-rat-plugin/check-mojo.html) be configured with multiple exclusion files? I tried using multiple times the property 'excludesFile'  but it seems to retain only the last one provided.

**Comment by vlsi** [comment]
> Can the apache-rat-plugin be configured with multiple exclusion files?

The current build script uses two exclude files: https://github.com/apache/calcite-avatica/blob/af451b93d9f4edbe205eb0d9885d9af05c49fde9/build.gradle.kts#L105-L108

`.gitignore` (which is the same ignore file that is used by Git), and `.ratignore`

**Comment by F21** [comment]
@vlsi Thanks for working on this! I pulled in your latest commits and achieve my first successful build in a docker container. Speed was also quite good:

```
BUILD SUCCESSFUL in 4m 21s
104 actionable tasks: 64 executed, 40 up-to-date
```

The reason I am using a docker container for making a release is because I usually don't have a Java environment set up on my machine (which is on Windows as well). Docker containers also allow me to swap between Java/JDK versions easily and there is usually nothing to install as the image is already built.

I did have to use the JDK flavor instead of the JRE flavor of the Gradle container, otherwise it complains about JDK being missing.

These are the commands I used to build:
```
docker run -v "$PWD":/home/gradle/project:rw -v "/c/Users/Francis Chuang/AppData/Roaming/gnupg":/.gnupg -w /home/gradle/project --network="host" -it gradle:5.6-jdk sh
apt update
apt install gnupg2
mkdir -p /root/.gnupg
cp -r /.gnupg/ /root/
chmod -R 700 /root/.gnupg/
rm -rf /root/.gnupg/*.lock
export GPG_TTY=/dev/console
touch /root/.gnupg/gpg-agent.conf
echo 'default-cache-ttl 10000' >> /root/.gnupg/gpg-agent.conf
echo 'max-cache-ttl 10000' >> /root/.gnupg/gpg-agent.conf
"test" | gpg2 --local-user 635665E0BE3F72552910CB74BBE44E923A970AB7 --output /dev/null --sign -
gradle prepareVote -PasfTestSvnUsername=test -PasfTestSvnPassword=test -PasfTestNexusUsername=test -PasfTestNexusPassword=test -PasfTestGitSourceUsername=test -PasfTestGitSourcePassword=test -PallowUncommittedChanges -Prc=0 -PuseGpgCmd -Psigning.gnupg.keyName=635665E0BE3F72552910CB74BBE44E923A970AB7
```

I agree that the `docker.sh` file can be massively simplified. One of my initial motivations with the script is to make it interactive. This is so that the release manager will not need to mess around with command line arguments and properties. I think for releasing using docker, the interactive aspects in the script to select the GPG key and set the appropriate command line parameters should be kept, but the redundant parts that are now being managed within the gradle tasks should be removed.

I have only tried the `prepareVote` task so far. Are there any other ones I should try?

**Comment by vlsi** [comment]
> Speed was also quite good

Just in case, you can add `-m` (== `--dry-run`) and it will print the list of tasks it is about to execute.

For instance, `prepareVote` does not execute tests.

>I have only tried the prepareVote task so far. Are there any other ones I should try?

`./gradlew publishDist -Prc=0` to publish the candidate.

> This is so that the release manager will not need to mess around with command line arguments and properties

Most of the properties are a one-time configuration for the machine.
For instance, `signing.gnupg.keyName`, `asfCommitterId`, `asf*Username` could be saved to the properties in the home directory.
`asfTest*` could be saved as well.

I do not think it would be great to type the same `asfCommitterId` again and again in the "interactive" mode.

What if the script takes certain properties from `$HOME/.gradle/gradle.properties`?
Then it would reduce the set of properties to type.

> the interactive aspects in the script to select the GPG key and set the appropriate command

It is sad that "release with Docker" requires lots of commands to type.
To release "without Docker" one needs "Java, Git, SVN" (ps. SVN seems to be missing in the docs).
The release itself is 2 short commands, and the script would suggest the missing properties automatically.

The release "with Docker" requires significant preparations (gpg install, copy keyring, etc, etc).
It could probably make sense to add a Dockerfile with gpg installed so one could just run a simple command and get suitable environment for the release / verification.

**Comment by F21** [comment]
> What if the script takes certain properties from $HOME/.gradle/gradle.properties?

I am not sure how easily this can be mounted into a docker container via docker-compose as `$HOME` is an environment variable. We also need to contend with Windows machines.

> I do not think it would be great to type the same asfCommitterId again and again in the "interactive" mode.
I haven't found this to be an issue when making the last couple of releases for Avatica. I've found the interactive mode to be easier as the script asks for everything up front (the asf username, etc is only asked once and then used when it's required within a given task) and then everything else is automated. The releases has been smoother and less error-prone from past experience.

> The release "with Docker" requires significant preparations (gpg install, copy keyring, etc, etc).
It could probably make sense to add a Dockerfile with gpg installed so one could just run a simple command and get suitable environment for the release / verification.

This is all automated in `docker.sh`, so it hasn't really been a problem in my experience.

**Comment by zabetak** [comment]
> The current build script uses two exclude files:

This is partially true :) Indeed the script uses the two files but not through `excludeFile` property. If I understand well the script reads the lines of .ratignore file and creates one `exclude` entry for each. This is possible thanks to Kotlin and gradle but I guess it is not possible to do with a maven build.

**Comment by joshelser** [comment]
@vlsi please update https://calcite.apache.org/avatica/docs/howto.html

This should have been done prior to merging in the change so that everyone else can know how to build :\

**Comment by vlsi** [comment]
>This should have been done prior to merging in the change

Do you suggest I should have updated the site with non-committed changes?
The doc was updated in this PR: https://github.com/apache/calcite-avatica/blob/master/site/_docs/howto.md


In any case, it looks like I've not enough grants to update the site:

```
$ svn commit -m "Update avatica/docs/ to represent Maven -> Gradle" howto.html
Sending        howto.html
Transmitting file data .svn: E195023: Commit failed (details follow):
svn: E195023: Changing file '/Users/vladimirsitnikov/Documents/code/calcite-avatica/site/target/avatica/docs/howto.html' is forbidden by the server
svn: E175013: Access to '/repos/asf/!svn/txr/1869637-151qa/calcite/site/avatica/docs/howto.html' forbidden
```

**Comment by vlsi** [comment]
Holy cow. The website has been migrated to Git already, however "howto" is still suggesting `svn` which I tried to execute.

**Comment by joshelser** [comment]
Having the site changes staged to roll out at the same time would be my preferred stance. The worry is that an author forgets to make the site changes and then we linger.

However, seems like we were already there if the site didn't reflect svn->git for the website 😑.

Nonetheless, thanks for picking it up, @vlsi

**Comment by vlsi** [comment]
Just in case, I think Gradle is pretty good with integrated help.

For instance:

```
$ ./gradlew
...
To run a build, run gradlew <task> ...

To see a list of available tasks, run gradlew tasks

To see a list of command-line options, run gradlew --help

To see more detail about a task, run gradlew help --task <task>
```

```
$ ./gradlew tasks

...
Build tasks
-----------
assemble - Assembles the outputs of this project.
build - Assembles and tests this project.
buildDependents - Assembles and tests this project and all projects that depend on it.
buildNeeded - Assembles and tests this project and all projects it depends on.
classes - Assembles main classes.
clean - Deletes the build directory.
generatePom - Generate pom.xml files (see build/publications/core/pom*.xml)
jar - Assembles a jar archive containing the main classes.
license - Generates LICENSE file for the source distribution
testClasses - Assembles test classes
...
Release tasks
-------------
prepareVote - Stage artifacts and prepare text for vote mail
publishDist - Publish release artifacts to SVN and Nexus
removeStaleArtifacts - Removes stale artifacts from dist.apache.org (dry run with -PasfDryRun)
...
Verification tasks
------------------
check - Runs all checks.
forbiddenApis - Runs forbidden-apis checks.
rat - Runs Apache Rat audit tool
test - Runs the unit tests.
...
```

It is not perfect, but it is quite good, and I think howto should not try to duplicate that.
In other words, I would invest time in perfecting `./gradlew tasks` output rather than in updating basic instructions in `howto.md`.

**Comment by F21** [comment]
Thanks for getting this in @vlsi !

**Comment by joshelser** [comment]
Thanks for the pointer on the `tasks` command. I'm coming from net-zero :)
